README.md
# openapi-java-client

Simple emlpoyee contract tracker api
- API version: 1.0.0
  - Build date: 2023-03-24T09:34:46.789412-07:00[America/Los_Angeles]

This API allows management of client contracts and tracks history of employee contract assignments



*Automatically generated by the [OpenAPI Generator](https://openapi-generator.tech)*


## Requirements

Building the API client library requires:
1. Java 1.8+
2. Maven (3.8.3+)/Gradle (7.2+)

## Installation

To install the API client library to your local Maven repository, simply execute:

```shell
mvn clean install
```

To deploy it to a remote Maven repository instead, configure the settings of the repository and execute:

```shell
mvn clean deploy
```

Refer to the [OSSRH Guide](http://central.sonatype.org/pages/ossrh-guide.html) for more information.

### Maven users

Add this dependency to your project's POM:

```xml
<dependency>
  <groupId>com.boxboat</groupId>
  <artifactId>openapi-java-client</artifactId>
  <version>1.0.0</version>
  <scope>compile</scope>
</dependency>
```

### Gradle users

Add this dependency to your project's build file:

```groovy
  repositories {
    mavenCentral()     // Needed if the 'openapi-java-client' jar has been published to maven central.
    mavenLocal()       // Needed if the 'openapi-java-client' jar has been published to the local maven repo.
  }

  dependencies {
     implementation "com.boxboat:openapi-java-client:1.0.0"
  }
```

### Others

At first generate the JAR by executing:

```shell
mvn clean package
```

Then manually install the following JARs:

* `target/openapi-java-client-1.0.0.jar`
* `target/lib/*.jar`

## Getting Started

Please follow the [installation](#installation) instruction and execute the following Java code:

```java

// Import classes:
import com.boxboat.simpletracker.api.ApiClient;
import com.boxboat.simpletracker.api.ApiException;
import com.boxboat.simpletracker.api.Configuration;
import com.boxboat.simpletracker.api.models.*;
import com.boxboat.simpletracker.api.DefaultApi;

public class Example {
  public static void main(String[] args) {
    ApiClient defaultClient = Configuration.getDefaultApiClient();
    defaultClient.setBasePath("http://localhost");

    DefaultApi apiInstance = new DefaultApi(defaultClient);
    Integer clientId = 56; // Integer | Numeric ID of the client
    try {
      apiInstance.clientsClientIdDelete(clientId);
    } catch (ApiException e) {
      System.err.println("Exception when calling DefaultApi#clientsClientIdDelete");
      System.err.println("Status code: " + e.getCode());
      System.err.println("Reason: " + e.getResponseBody());
      System.err.println("Response headers: " + e.getResponseHeaders());
      e.printStackTrace();
    }
  }
}

```

## Documentation for API Endpoints

All URIs are relative to *http://localhost*

Class | Method | HTTP request | Description
------------ | ------------- | ------------- | -------------
*DefaultApi* | [**clientsClientIdDelete**](docs/DefaultApi.md#clientsClientIdDelete) | **DELETE** /clients/{clientId} | Delete a client by ID
*DefaultApi* | [**clientsClientIdGet**](docs/DefaultApi.md#clientsClientIdGet) | **GET** /clients/{clientId} | Get a client by ID
*DefaultApi* | [**clientsClientIdPut**](docs/DefaultApi.md#clientsClientIdPut) | **PUT** /clients/{clientId} | Update a client by ID
*DefaultApi* | [**clientsGet**](docs/DefaultApi.md#clientsGet) | **GET** /clients | Get all clients
*DefaultApi* | [**clientsPost**](docs/DefaultApi.md#clientsPost) | **POST** /clients | Create new client
*DefaultApi* | [**contractsContractIdDelete**](docs/DefaultApi.md#contractsContractIdDelete) | **DELETE** /contracts/{contractId} | Delete a contract by ID
*DefaultApi* | [**contractsContractIdGet**](docs/DefaultApi.md#contractsContractIdGet) | **GET** /contracts/{contractId} | Get a contract by ID
*DefaultApi* | [**contractsContractIdPut**](docs/DefaultApi.md#contractsContractIdPut) | **PUT** /contracts/{contractId} | Update a contract by ID
*DefaultApi* | [**contractsGet**](docs/DefaultApi.md#contractsGet) | **GET** /contracts | Get all contracts
*DefaultApi* | [**contractsPost**](docs/DefaultApi.md#contractsPost) | **POST** /contracts | Create new contract
*DefaultApi* | [**employeesEmployeeIdDelete**](docs/DefaultApi.md#employeesEmployeeIdDelete) | **DELETE** /employees/{employeeId} | Delete an employee by ID
*DefaultApi* | [**employeesEmployeeIdGet**](docs/DefaultApi.md#employeesEmployeeIdGet) | **GET** /employees/{employeeId} | Get an employee by ID
*DefaultApi* | [**employeesEmployeeIdPut**](docs/DefaultApi.md#employeesEmployeeIdPut) | **PUT** /employees/{employeeId} | Update an employee by ID
*DefaultApi* | [**employeesGet**](docs/DefaultApi.md#employeesGet) | **GET** /employees | Get all employees
*DefaultApi* | [**employeesPost**](docs/DefaultApi.md#employeesPost) | **POST** /employees | Create new employee
*DefaultApi* | [**historyGet**](docs/DefaultApi.md#historyGet) | **GET** /history | Get all history
*DefaultApi* | [**historyHistoryIdDelete**](docs/DefaultApi.md#historyHistoryIdDelete) | **DELETE** /history/{historyId} | Delete an employee contract assignment by ID
*DefaultApi* | [**historyHistoryIdGet**](docs/DefaultApi.md#historyHistoryIdGet) | **GET** /history/{historyId} | Get an employee contract asignment by ID
*DefaultApi* | [**historyHistoryIdPut**](docs/DefaultApi.md#historyHistoryIdPut) | **PUT** /history/{historyId} | Update an employee contract assignment by ID
*DefaultApi* | [**historyPost**](docs/DefaultApi.md#historyPost) | **POST** /history | Create new employee contract assignment


## Documentation for Models

 - [Client](docs/Client.md)
 - [ClientsPostRequest](docs/ClientsPostRequest.md)
 - [Contract](docs/Contract.md)
 - [ContractPatch](docs/ContractPatch.md)
 - [Employee](docs/Employee.md)
 - [EmployeesPostRequest](docs/EmployeesPostRequest.md)
 - [History](docs/History.md)
 - [HistoryPatch](docs/HistoryPatch.md)


## Documentation for Authorization

All endpoints do not require authorization.
Authentication schemes defined for the API:

## Recommendation

It's recommended to create an instance of `ApiClient` per thread in a multithreaded environment to avoid any potential issues.

## Author





docs/History.md


# History

employee contract assignment data

## Properties

| Name | Type | Description | Notes |
|------------ | ------------- | ------------- | -------------|
|**id** | **Integer** |  |  [optional] |
|**clientId** | **Integer** |  |  [optional] |
|**contractId** | **Integer** |  |  [optional] |
|**employeeId** | **Integer** |  |  [optional] |
|**clientName** | **String** |  |  [optional] |
|**employeeName** | **String** |  |  [optional] |
|**role** | **String** |  |  [optional] |





src/main/java/com/boxboat/simpletracker/model/History.java
/*
 * Simple emlpoyee contract tracker api
 * This API allows management of client contracts and tracks history of employee contract assignments 
 *
 * The version of the OpenAPI document: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


package com.boxboat.simpletracker.model;

import java.util.Objects;
import java.util.Arrays;
import com.google.gson.TypeAdapter;
import com.google.gson.annotations.JsonAdapter;
import com.google.gson.annotations.SerializedName;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import java.io.IOException;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.TypeAdapterFactory;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

import com.boxboat.simpletracker.api.JSON;

/**
 * employee contract assignment data
 */
@javax.annotation.Generated(value = "org.openapitools.codegen.languages.JavaClientCodegen", date = "2023-03-24T09:34:46.789412-07:00[America/Los_Angeles]")
public class History {
  public static final String SERIALIZED_NAME_ID = "id";
  @SerializedName(SERIALIZED_NAME_ID)
  private Integer id;

  public static final String SERIALIZED_NAME_CLIENT_ID = "clientId";
  @SerializedName(SERIALIZED_NAME_CLIENT_ID)
  private Integer clientId;

  public static final String SERIALIZED_NAME_CONTRACT_ID = "contractId";
  @SerializedName(SERIALIZED_NAME_CONTRACT_ID)
  private Integer contractId;

  public static final String SERIALIZED_NAME_EMPLOYEE_ID = "employeeId";
  @SerializedName(SERIALIZED_NAME_EMPLOYEE_ID)
  private Integer employeeId;

  public static final String SERIALIZED_NAME_CLIENT_NAME = "clientName";
  @SerializedName(SERIALIZED_NAME_CLIENT_NAME)
  private String clientName;

  public static final String SERIALIZED_NAME_EMPLOYEE_NAME = "employeeName";
  @SerializedName(SERIALIZED_NAME_EMPLOYEE_NAME)
  private String employeeName;

  public static final String SERIALIZED_NAME_ROLE = "role";
  @SerializedName(SERIALIZED_NAME_ROLE)
  private String role;

  public History() {
  }

  public History id(Integer id) {
    
    this.id = id;
    return this;
  }

   /**
   * Get id
   * @return id
  **/
  @javax.annotation.Nullable

  public Integer getId() {
    return id;
  }


  public void setId(Integer id) {
    this.id = id;
  }


  public History clientId(Integer clientId) {
    
    this.clientId = clientId;
    return this;
  }

   /**
   * Get clientId
   * @return clientId
  **/
  @javax.annotation.Nullable

  public Integer getClientId() {
    return clientId;
  }


  public void setClientId(Integer clientId) {
    this.clientId = clientId;
  }


  public History contractId(Integer contractId) {
    
    this.contractId = contractId;
    return this;
  }

   /**
   * Get contractId
   * @return contractId
  **/
  @javax.annotation.Nullable

  public Integer getContractId() {
    return contractId;
  }


  public void setContractId(Integer contractId) {
    this.contractId = contractId;
  }


  public History employeeId(Integer employeeId) {
    
    this.employeeId = employeeId;
    return this;
  }

   /**
   * Get employeeId
   * @return employeeId
  **/
  @javax.annotation.Nullable

  public Integer getEmployeeId() {
    return employeeId;
  }


  public void setEmployeeId(Integer employeeId) {
    this.employeeId = employeeId;
  }


  public History clientName(String clientName) {
    
    this.clientName = clientName;
    return this;
  }

   /**
   * Get clientName
   * @return clientName
  **/
  @javax.annotation.Nullable

  public String getClientName() {
    return clientName;
  }


  public void setClientName(String clientName) {
    this.clientName = clientName;
  }


  public History employeeName(String employeeName) {
    
    this.employeeName = employeeName;
    return this;
  }

   /**
   * Get employeeName
   * @return employeeName
  **/
  @javax.annotation.Nullable

  public String getEmployeeName() {
    return employeeName;
  }


  public void setEmployeeName(String employeeName) {
    this.employeeName = employeeName;
  }


  public History role(String role) {
    
    this.role = role;
    return this;
  }

   /**
   * Get role
   * @return role
  **/
  @javax.annotation.Nullable

  public String getRole() {
    return role;
  }


  public void setRole(String role) {
    this.role = role;
  }



  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    History history = (History) o;
    return Objects.equals(this.id, history.id) &&
        Objects.equals(this.clientId, history.clientId) &&
        Objects.equals(this.contractId, history.contractId) &&
        Objects.equals(this.employeeId, history.employeeId) &&
        Objects.equals(this.clientName, history.clientName) &&
        Objects.equals(this.employeeName, history.employeeName) &&
        Objects.equals(this.role, history.role);
  }

  @Override
  public int hashCode() {
    return Objects.hash(id, clientId, contractId, employeeId, clientName, employeeName, role);
  }

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class History {\n");
    sb.append("    id: ").append(toIndentedString(id)).append("\n");
    sb.append("    clientId: ").append(toIndentedString(clientId)).append("\n");
    sb.append("    contractId: ").append(toIndentedString(contractId)).append("\n");
    sb.append("    employeeId: ").append(toIndentedString(employeeId)).append("\n");
    sb.append("    clientName: ").append(toIndentedString(clientName)).append("\n");
    sb.append("    employeeName: ").append(toIndentedString(employeeName)).append("\n");
    sb.append("    role: ").append(toIndentedString(role)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  private String toIndentedString(Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }


  public static HashSet<String> openapiFields;
  public static HashSet<String> openapiRequiredFields;

  static {
    // a set of all properties/fields (JSON key names)
    openapiFields = new HashSet<String>();
    openapiFields.add("id");
    openapiFields.add("clientId");
    openapiFields.add("contractId");
    openapiFields.add("employeeId");
    openapiFields.add("clientName");
    openapiFields.add("employeeName");
    openapiFields.add("role");

    // a set of required properties/fields (JSON key names)
    openapiRequiredFields = new HashSet<String>();
  }

 /**
  * Validates the JSON Object and throws an exception if issues found
  *
  * @param jsonObj JSON Object
  * @throws IOException if the JSON Object is invalid with respect to History
  */
  public static void validateJsonObject(JsonObject jsonObj) throws IOException {
      if (jsonObj == null) {
        if (!History.openapiRequiredFields.isEmpty()) { // has required fields but JSON object is null
          throw new IllegalArgumentException(String.format("The required field(s) %s in History is not found in the empty JSON string", History.openapiRequiredFields.toString()));
        }
      }

      Set<Entry<String, JsonElement>> entries = jsonObj.entrySet();
      // check to see if the JSON string contains additional fields
      for (Entry<String, JsonElement> entry : entries) {
        if (!History.openapiFields.contains(entry.getKey())) {
          throw new IllegalArgumentException(String.format("The field `%s` in the JSON string is not defined in the `History` properties. JSON: %s", entry.getKey(), jsonObj.toString()));
        }
      }
      if ((jsonObj.get("clientName") != null && !jsonObj.get("clientName").isJsonNull()) && !jsonObj.get("clientName").isJsonPrimitive()) {
        throw new IllegalArgumentException(String.format("Expected the field `clientName` to be a primitive type in the JSON string but got `%s`", jsonObj.get("clientName").toString()));
      }
      if ((jsonObj.get("employeeName") != null && !jsonObj.get("employeeName").isJsonNull()) && !jsonObj.get("employeeName").isJsonPrimitive()) {
        throw new IllegalArgumentException(String.format("Expected the field `employeeName` to be a primitive type in the JSON string but got `%s`", jsonObj.get("employeeName").toString()));
      }
      if ((jsonObj.get("role") != null && !jsonObj.get("role").isJsonNull()) && !jsonObj.get("role").isJsonPrimitive()) {
        throw new IllegalArgumentException(String.format("Expected the field `role` to be a primitive type in the JSON string but got `%s`", jsonObj.get("role").toString()));
      }
  }

  public static class CustomTypeAdapterFactory implements TypeAdapterFactory {
    @SuppressWarnings("unchecked")
    @Override
    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
       if (!History.class.isAssignableFrom(type.getRawType())) {
         return null; // this class only serializes 'History' and its subtypes
       }
       final TypeAdapter<JsonElement> elementAdapter = gson.getAdapter(JsonElement.class);
       final TypeAdapter<History> thisAdapter
                        = gson.getDelegateAdapter(this, TypeToken.get(History.class));

       return (TypeAdapter<T>) new TypeAdapter<History>() {
           @Override
           public void write(JsonWriter out, History value) throws IOException {
             JsonObject obj = thisAdapter.toJsonTree(value).getAsJsonObject();
             elementAdapter.write(out, obj);
           }

           @Override
           public History read(JsonReader in) throws IOException {
             JsonObject jsonObj = elementAdapter.read(in).getAsJsonObject();
             validateJsonObject(jsonObj);
             return thisAdapter.fromJsonTree(jsonObj);
           }

       }.nullSafe();
    }
  }

 /**
  * Create an instance of History given an JSON string
  *
  * @param jsonString JSON string
  * @return An instance of History
  * @throws IOException if the JSON string is invalid with respect to History
  */
  public static History fromJson(String jsonString) throws IOException {
    return JSON.getGson().fromJson(jsonString, History.class);
  }

 /**
  * Convert an instance of History to an JSON string
  *
  * @return JSON string
  */
  public String toJson() {
    return JSON.getGson().toJson(this);
  }
}

